contracts/lib/OnChainIdentity.sol

```diff
@@ -0,0 +1,272 @@
+        return self.trees.claimsTree.getProofByRoot(claimIndexHash, root);
+    }
+
+    /**
+     * @dev Retrieve ClaimsTree latest root.
+     * @return The latest ClaimsTree root
+     */
+    function getClaimsTreeRoot(Identity storage self) external view returns (uint256) {
+        return self.trees.claimsTree.getRoot();
+    }
+
+    /**
+     * @dev Retrieve inclusion or non-inclusion proof for a given revocation nonce.
+     * @param revocationNonce - revocation nonce
+     * @return The RevocationsTree inclusion or non-inclusion proof for the claim
+     */
+    function getRevocationProof(Identity storage self, uint64 revocationNonce) external view returns (SmtLib.Proof memory) {
+        return self.trees.revocationsTree.getProof(uint256(revocationNonce));
+    }
+
+    /**
+     * @dev Retrieve inclusion or non-inclusion proof for a given revocation nonce by target root.
+     * @param revocationNonce - revocation nonce
+     * @param root - root of the tree
+     * @return The RevocationsTree inclusion or non-inclusion proof for the claim
+     */
+    function getRevocationProofByRoot(Identity storage self, uint64 revocationNonce, uint256 root) external view returns (SmtLib.Proof memory) {
+        return self.trees.revocationsTree.getProofByRoot(uint256(revocationNonce), root);
+    }
+
+    /**
+     * @dev Retrieve RevocationsTree latest root.
+     * @return The latest RevocationsTree root
+     */
+    function getRevocationsTreeRoot(Identity storage self) external view returns (uint256) {
+        return self.trees.revocationsTree.getRoot();
+    }
+
+    /**
+     * @dev Retrieve inclusion or non-inclusion proof for a given claimsTreeRoot.
+     * @param claimsTreeRoot - claims tree root
+     * @return The RevocationsTree inclusion or non-inclusion proof for the claim
+     */
+    function getRootProof(Identity storage self, uint256 claimsTreeRoot) external view returns (SmtLib.Proof memory) {
+        return self.trees.rootsTree.getProof(claimsTreeRoot);
+    }
+
+    /**
+     * @dev Retrieve inclusion or non-inclusion proof for a given claimsTreeRoot by target root.
+     * @param claimsTreeRoot - claims tree root
+     * @param root - root of the tree
+     * @return The RevocationsTree inclusion or non-inclusion proof for the claim
+     */
+    function getRootProofByRoot(Identity storage self, uint256 claimsTreeRoot, uint256 root) external view returns (SmtLib.Proof memory) {
+        return self.trees.rootsTree.getProofByRoot(claimsTreeRoot, root);
+    }
+
+    /**
+     * @dev Retrieve RootsTree latest root.
+     * @return The latest RootsTree root
+     */
+    function getRootsTreeRoot(Identity storage self) external view returns (uint256) {
+        return self.trees.rootsTree.getRoot();
+    }
+
+       /**
+     * @dev write roots to history by state
+     * @param self identity
+     * @param roots set of roots
+     */
+    function writeHistory(Identity storage self, Roots memory roots) internal {
```

Comment:
Please pass latestState as a separate param `state`, it could be useful in the future to write history not from the current state of trees.